Include source location in schema entity types and action types

eidelmanjonathan opened the pull request:
## Description of changes

Include source locations in entity types and action types in schema. This allows us to give more informative error messages when validating policies and can be useful as part of an LSP/IDE plugin

## Issue #, if available

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ x ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [ x ] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [ x ] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [ x ] Does not require updates because my change does not impact the Cedar language specification.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in [`cedar-docs`](https://github.com/cedar-policy/cedar-docs). PRs should be targeted at a `staging-X.Y` branch, not `main`.)
- [ ] I'm not sure how my change impacts the documentation. (Post your PR anyways, and we'll discuss in the comments.)

cdisselkoen reviewed (approved):
Looks great; two extremely minor nits

eidelmanjonathan commented on cedar-policy-validator/src/cedar_schema/to_json_schema.rs:
```diff
@@ -236,15 +236,20 @@ fn convert_action_decl(
-        loc: Some(a.data.loc),
-    };
+
     // Then map that type across all of the bound names
```
Good point, will fix

eidelmanjonathan commented on cedar-policy/src/api.rs:
```diff
@@ -30,7 +30,7 @@ use cedar_policy_validator::entity_manifest;
     AccessTrie, EntityManifest, EntityRoot, Fields, RootAccessTrie,
 };
-use cedar_policy_validator::json_schema;
+use cedar_policy_validator::json_schema::{self};
```
Good point, will fix

john-h-kastner-aws reviewed (commented):
Is there a reason this needs the `extended-schema` feature? It does makes sense as a way to be absolutely certain we're not changing the JSON schema format, but I'm wondering if there's some other reason we need it.

eidelmanjonathan commented:
> Is there a reason this needs the `extended-schema` feature? It does makes sense as a way to be absolutely certain we're not changing the JSON schema format, but I'm wondering if there's some other reason we need it.

I added in the `extended-schema` because we were failing the DRT build because I added the `defn_loc` to the `ActionType`. That said, there will be a subsequent, more substantial PR coming after this which will store the locations of common types in the `ValidatorSchema` and I think the feature flag is definitely justified for that. So overall I think it is worth having.

john-h-kastner-aws reviewed (approved):
Looks fine
